src/verifier/builder_visitor.rs

```diff
@@ -243,21 +284,55 @@ impl SymbolTableBuilder<'_> {
     }
 
     fn add_fn(&mut self, f: &mut Fn) {
-        let f_name = &f.name;
-        if let Some(other_fn_id) = self.table.lookup(&f_name.name) {
+        let f_id: &parser_types::FnId = &f.name;
+        //TODO - make work with if self.check_duplicate_id(&f_id.name, f_id.loc, f.is_comp_proivded){ return; }
+
+        if let Some(other_fn_id) = self.table.lookup(&f_id.name) {
             if let Some(other_rec) = self.table.get_record(other_fn_id) {
-                if let (Some(curr_loc), Some(other_loc)) = (&f_name.loc, other_rec.loc()) {
+                if let (Some(curr_loc), Some(other_loc)) = (&f_id.loc, other_rec.loc()) {
                     self.err.duplicate_identifier_error(
                         false,
-                        f_name.name.clone(),
+                        f_id.name.clone(),
                         Some(curr_loc.line_col.clone()),
                         Some(other_loc.line_col.clone()),
                     );
                 } else {
-                    // This should never be the case since it's controlled by the compiler!
-                    self.err
-                        .unexpected_error(true, Some(UNEXPECTED_ERR_MSG.to_string()), None);
-                    unreachable!()
+                    // If there is another fn with the same name as a compiler generated fn, throw a duplicate id error
+                    match &f_id.loc {
+                        Some(loc) => {
+                            //add check if the record "other_rec" is a compiler provided function
+                            match other_rec {
+                                Record::Fn {
+                                    is_comp_provided, ..
+                                } => {
+                                    if *is_comp_provided {
+                                        self.err.compiler_fn_overload_error(
+                                            false,
+                                            f_id.name.clone(),
+                                            Some(loc.line_col.clone()),
+                                        );
+                                    } else {
+                                        self.err.unexpected_error(
+                                            true,
+                                            Some(UNEXPECTED_ERR_MSG.to_string()),
+                                            None,
+                                        );
+                                    }
+                                }
+                                _ => {
+                                    self.err.unexpected_error(
+                                        true,
+                                        Some(UNEXPECTED_ERR_MSG.to_string()),
+                                        None,
+                                    );
+                                }
```

Comment:
Wouldn't this be a normal dup ID error? (two user-defined symbols with the same name?)